# 套件

In [1]:
import random
import numpy as np
import pandas as pd

# 參數

In [2]:
from datetime import datetime
#y_symbol = "^TWII" # 股票代碼 ^TWII 台灣加權指數 美股蘋果 AAPL 台股台積電 2330.TW

y_twii = "^TWII"
y_2330 = "2330.TW"

startdate = datetime(2019,1,1) # 訓練起始日期
enddate = datetime(2020,1,1) # 訓練結束日期

# 載入資料

In [3]:
import yfinance as yf
from datetime import datetime, timedelta

#df = yf.download(y_symbol, start = startdate,end = enddate)

df_TWII = yf.download(y_twii, start = startdate,end = enddate)

df_2330 =  yf.download(y_2330, start = startdate,end = enddate)

display(df_TWII.head(20))
display(df_2330)

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


Open          High          Low         Close  \
Date                                                                
2019-01-02   9725.269531   9737.679688  9542.519531   9554.139648   
2019-01-03   9494.250000   9554.530273  9474.469727   9492.419922   
2019-01-04   9427.240234   9427.240234  9319.280273   9382.509766   
2019-01-07   9460.200195   9596.349609  9460.200195   9590.299805   
2019-01-08   9591.559570   9600.940430  9547.629883   9563.599609   
2019-01-09   9584.629883   9741.639648  9584.629883   9738.309570   
2019-01-10   9731.110352   9731.110352  9683.200195   9720.690430   
2019-01-11   9753.629883   9796.980469  9736.000000   9759.400391   
2019-01-14   9737.669922   9762.370117  9677.490234   9708.219727   
2019-01-15   9701.929688   9806.040039  9701.929688   9806.040039   
2019-01-16   9781.349609   9803.990234  9763.809570   9763.809570   
2019-01-17   9773.309570   9826.250000  9759.509766   9789.150391   
2019-01-18   9781.389648   9842.799805  9771.150391   9836.059570   
2019-01-21   9870.269531   9919.940430  9870.269531   9889.400391   
2019-01-22   9880.540039   9894.660156  9855.469727   9894.660156   
2019-01-23   9869.540039   9873.389648  9828.030273   9846.400391   
2019-01-24   9865.000000   9892.820312  9847.320312   9877.120117   
2019-01-25   9923.389648   9980.080078  9918.209961   9969.610352   
2019-01-28  10006.459961  10020.599609  9990.070312  10013.330078   
2019-01-29   9944.709961   9952.339844  9893.959961   9931.589844   

               Adj Close   Volume  
Date                               
2019-01-02   9554.139648  1239100  
2019-01-03   9492.419922  1560800  
2019-01-04   9382.509766  1842100  
2019-01-07   9590.299805  1722300  
2019-01-08   9563.599609  1423100  
2019-01-09   9738.309570  2281800  
2019-01-10   9720.690430  1609200  
2019-01-11   9759.400391  1618600  
2019-01-14   9708.219727  1356000  
2019-01-15   9806.040039  1709500  
2019-01-16   9763.809570  1796800  
2019-01-17   9789.150391  1634200  
2019-01-18   9836.059570  1545600  
2019-01-21   9889.400391  1593600  
2019-01-22   9894.660156  1408000  
2019-01-23   9846.400391  1140300  
2019-01-24   9877.120117  1452000  
2019-01-25   9969.610352  1944600  
2019-01-28  10013.330078  1656900  
2019-01-29   9931.589844  1647200

Open   High    Low  Close   Adj Close    Volume
Date                                                        
2019-01-02  226.5  226.5  219.0  219.5  193.747650  32900482
2019-01-03  214.0  218.0  214.0  215.5  190.216919  34087620
2019-01-04  211.5  211.5  206.5  208.0  183.596832  65943521
2019-01-07  212.0  214.0  211.0  213.0  188.010223  35442176
2019-01-08  212.0  212.5  210.0  211.0  186.244843  22694481
...           ...    ...    ...    ...         ...       ...
2019-12-25  332.0  334.5  331.0  333.0  308.237762  12262954
2019-12-26  333.0  334.0  331.5  333.0  308.237762  11427884
2019-12-27  335.0  338.0  335.0  338.0  312.865967  16677806
2019-12-30  338.0  339.0  334.0  334.5  309.626251  21531263
2019-12-31  331.0  333.5  331.0  331.0  306.386536  22890517

[241 rows x 6 columns]

# 算出「強勢股」「弱勢股」
* 把台股加權指數當作基準指數(benchmark index)，以個股作為比較指數，比較同一個時期內的累積漲跌幅，就可以算出個股是否比大盤強或弱。

## 步驟一、先計算出台股加權指數20天的漲跌幅%

2019/3/4 台股加權指數收盤是10349.88點，3/29加權指數收盤是10541.43點
加權指數20天的漲跌幅%= ((10541.43-10349.88)/ 10349.88)x100= + 1.85%

In [4]:
print(df_TWII['Adj Close'])

# print(df.loc[:,'Adj Close'])


Date
2019-01-02     9554.139648
2019-01-03     9492.419922
2019-01-04     9382.509766
2019-01-07     9590.299805
2019-01-08     9563.599609
                  ...     
2019-12-25    12008.129883
2019-12-26    12001.009766
2019-12-27    12091.589844
2019-12-30    12053.370117
2019-12-31    11997.139648
Name: Adj Close, Length: 241, dtype: float64


In [5]:
# 假設你已經有一個名為 twii_close 的列表，包含了每天的收盤價
twii_close = df_TWII['Close']

window_size = 20  # 滾動窗口大小，即 20 天
twii_start_prices = []  # 用於存放每個滾動窗口的起始價格
twii_end_prices = []    # 用於存放每個滾動窗口的結束價格
twii_20days_returns = [] # 用於存放每個滾動窗口的漲跌幅

for i in range(len(twii_close) - window_size + 1):
    # 提取每個滾動窗口的起始價格和結束價格
    twii_start_price = twii_close[i]
    twii_end_price = twii_close[i + window_size - 1]
    
    twii_start_prices.append(twii_start_price)
    twii_end_prices.append(twii_end_price)

# 顯示結果
for start_price, end_price in zip(twii_start_prices, twii_end_prices):
#     print(f"Start Price: {start_price}, End Price: {end_price}")
    
    twii_20days_return = ((end_price - start_price) / start_price) * 100
    twii_20days_returns.append(twii_20days_return)
    print("台股加權指數 20 天的漲跌幅%:", twii_20days_return)

台股加權指數 20 天的漲跌幅%: 3.9506455756508525
台股加權指數 20 天的漲跌幅%: 4.633590247481595
台股加權指數 20 天的漲跌幅%: 6.626587660509446
台股加權指數 20 天的漲跌幅%: 5.291184217614091
台股加權指數 20 天的漲跌幅%: 5.51027322634264
台股加權指數 20 天的漲跌幅%: 3.601243036898509
台股加權指數 20 天的漲跌幅%: 3.5397675323465863
台股加權指數 20 天的漲跌幅%: 3.953930235131873
台股加權指數 20 天的漲跌幅%: 4.573856500667875
台股加權指數 20 天的漲跌幅%: 4.756455409288659
台股加權指數 20 天的漲跌幅%: 5.691638075518239
台股加權指數 20 天的漲跌幅%: 5.452664531144473
台股加權指數 20 天的漲跌幅%: 5.641182967844422
台股加權指數 20 天的漲跌幅%: 5.077652782048647
台股加權指數 20 天的漲跌幅%: 4.997743811470281
台股加權指數 20 天的漲跌幅%: 5.113335556279788
台股加權指數 20 天的漲跌幅%: 4.334660744810425
台股加權指數 20 天的漲跌幅%: 3.8872134957788225
台股加權指數 20 天的漲跌幅%: 2.9795243644945946
台股加權指數 20 天的漲跌幅%: 3.122965820474205
台股加權指數 20 天的漲跌幅%: 3.2018947884664812
台股加權指數 20 天的漲跌幅%: 3.3893603031211734
台股加權指數 20 天的漲跌幅%: 2.7291262374413523
台股加權指數 20 天的漲跌幅%: 2.5575369354578656
台股加權指數 20 天的漲跌幅%: 3.4714057859602407
台股加權指數 20 天的漲跌幅%: 4.45036960277344
台股加權指數 20 天的漲跌幅%: 3.6178403077092787
台股加權指數 20 天的漲跌幅%: 3.

## 步驟二、再算出台積電(2330)20天的漲跌幅%

2019/3/4 台積電收盤價是235.5元，2019/3/29台積電收盤價是245.5元
台積電20天的漲跌幅%=((245.5-235.5)/235.5)x100= + 4.25%

In [6]:
# 假設你已經有一個名為 t2300_close 的列表，包含了每天的收盤價
t2300_close = df_2330['Close']

window_size = 20  # 滾動窗口大小，即 20 天
t2300_start_prices = []  # 用於存放每個滾動窗口的起始價格
t2300_end_prices = []    # 用於存放每個滾動窗口的結束價格
t2300_20days_returns = [] # 用於存放每個滾動窗口的漲跌幅

for i in range(len(twii_close) - window_size + 1):
    # 提取每個滾動窗口的起始價格和結束價格
    t2300_start_price = t2300_close[i]
    t2300_end_price = t2300_close[i + window_size - 1]
    
    t2300_start_prices.append(t2300_start_price)
    t2300_end_prices.append(t2300_end_price)

# 顯示結果
for start_price, end_price in zip(t2300_start_prices, t2300_end_prices):
#     print(f"Start Price: {start_price}, End Price: {end_price}")
    
    t2300_20days_return = ((end_price - start_price) / start_price) * 100
    t2300_20days_returns.append(t2300_20days_return)
    print("台積電(2330) 20 天的漲跌幅%:", t2300_20days_return)

台積電(2330) 20 天的漲跌幅%: 1.366742596810934
台積電(2330) 20 天的漲跌幅%: 2.5522041763341066
台積電(2330) 20 天的漲跌幅%: 9.615384615384617
台積電(2330) 20 天的漲跌幅%: 7.981220657276995
台積電(2330) 20 天的漲跌幅%: 8.530805687203792
台積電(2330) 20 天的漲跌幅%: 5.336426914153132
台積電(2330) 20 天的漲跌幅%: 5.092592592592593
台積電(2330) 20 天的漲跌幅%: 4.308390022675737
台積電(2330) 20 天的漲跌幅%: 4.805491990846682
台積電(2330) 20 天的漲跌幅%: 6.108597285067873
台積電(2330) 20 天的漲跌幅%: 8.735632183908045
台積電(2330) 20 天的漲跌幅%: 7.2562358276643995
台積電(2330) 20 天的漲跌幅%: 8.924485125858125
台積電(2330) 20 天的漲跌幅%: 8.3710407239819
台積電(2330) 20 天的漲跌幅%: 7.174887892376682
台積電(2330) 20 天的漲跌幅%: 6.802721088435375
台積電(2330) 20 天的漲跌幅%: 4.719101123595506
台積電(2330) 20 天的漲跌幅%: 3.5398230088495577
台積電(2330) 20 天的漲跌幅%: 2.1834061135371177
台積電(2330) 20 天的漲跌幅%: 3.3707865168539324
台積電(2330) 20 天的漲跌幅%: 4.298642533936651
台積電(2330) 20 天的漲跌幅%: 3.289473684210526
台積電(2330) 20 天的漲跌幅%: 3.0434782608695654
台積電(2330) 20 天的漲跌幅%: 2.4017467248908297
台積電(2330) 20 天的漲跌幅%: 5.286343612334802
台積電(2330) 20 天的漲跌幅%:

## 步驟三、計算台積電強(弱)度值

(4.25 – 1.85)/1.85= + 1.297
由以上實例，我們可以算出台積電(2330)的強(弱)度值就是 + 1.297

In [7]:
strength_value = (t2300_20days_return - twii_20days_return) / twii_20days_return
print("台積電(2330) 的強(弱)度值:", strength_value)

台積電(2330) 的強(弱)度值: 0.9323114316460005


In [8]:
# 步驟三、計算台積電(2330) 的強(弱)度值
strength_values = [(t2300_return - twii_return) / twii_return for t2300_return, twii_return in zip(t2300_20days_returns, twii_20days_returns)]
print("台積電(2330) 的強(弱)度值:", strength_values)


台積電(2330) 的強(弱)度值: [-0.6540457576770174, -0.44919510789257716, 0.4510310748149666, 0.5083996944782051, 0.5481638272347492, 0.48182915162232764, 0.438679954560916, 0.0896474561929245, 0.05064336630259022, 0.28427510812751017, 0.534818635338587, 0.33076879867050446, 0.5820236955136912, 0.6486044011470377, 0.43562538678146234, 0.33038816122302384, 0.08868984250851532, -0.08936748323870063, -0.2671964225043414, 0.07935427751242484, 0.34253085061406646, -0.02947064046825129, 0.11518412710836297, -0.060914158621582296, 0.5228250277495355, 0.3858176806739346, 0.26186260393425803, 0.44334945881476856, 0.4294817907186277, 0.6386445068563965, 0.39398816944346365, 0.5567565226648957, -0.42588945023043456, -0.11341267808539206, 0.5094281820792802, 0.6387262990939764, 0.5942874129393058, 0.40269242724946797, 0.8327447608662222, 0.7379108870214749, 0.5482441759172525, 0.5077286443858388, 0.6912040589594515, 0.6516218670373881, 0.6811744845034272, 0.8928459976754504, 1.390083137179551, 1.287519823793

In [9]:
import random
import numpy as np
import pandas as pd
import yfinance as yf
from datetime import datetime, timedelta

y_twii = "^TWII"
y_2330 = "2330.TW"

startdate = datetime(2019, 1, 1)  # 訓練起始日期
enddate = datetime(2020, 1, 1)  # 訓練結束日期

df_TWII = yf.download(y_twii, start=startdate, end=enddate)
df_2330 = yf.download(y_2330, start=startdate, end=enddate)

# 步驟一、計算台股加權指數 20 天的漲跌幅%
twii_close = df_TWII['Close']
window_size = 20  # 滾動窗口大小，即 20 天
twii_start_prices = []  # 用於存放每個滾動窗口的起始價格
twii_end_prices = []    # 用於存放每個滾動窗口的結束價格

for i in range(len(twii_close) - window_size + 1):
    # 提取每個滾動窗口的起始價格和結束價格
    twii_start_price = twii_close[i]
    twii_end_price = twii_close[i + window_size - 1]
    
    twii_start_prices.append(twii_start_price)
    twii_end_prices.append(twii_end_price)

# 印出每個滾動窗口的結果
for start_price, end_price in zip(twii_start_prices, twii_end_prices):
    twii_20days_return = ((end_price - start_price) / start_price) * 100
#     print("台股加權指數 20 天的漲跌幅%:", twii_20days_return)

# 步驟二、計算台積電(2330) 20 天的漲跌幅%
t2300_close = df_2330['Close']
t2300_start_prices = []  # 用於存放每個滾動窗口的起始價格
t2300_end_prices = []    # 用於存放每個滾動窗口的結束價格

for i in range(len(t2300_close) - window_size + 1):
    # 提取每個滾動窗口的起始價格和結束價格
    t2300_start_price = t2300_close[i]
    t2300_end_price = t2300_close[i + window_size - 1]
    
    t2300_start_prices.append(t2300_start_price)
    t2300_end_prices.append(t2300_end_price)

# 印出每個滾動窗口的結果
for start_price, end_price in zip(t2300_start_prices, t2300_end_prices):
    t2300_20days_return = ((end_price - start_price) / start_price) * 100
#     print("台積電 20 天的漲跌幅%:", t2300_20days_return)

# 步驟三、計算台積電(2330) 的強(弱)度值
strength_values = [(t2300_20days_return - twii_return) / twii_return for t2300_20days_return, twii_return in zip(t2300_20days_returns, twii_20days_returns)]
print("台積電(2330) 的強(弱)度值:", strength_values)


[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
台積電(2330) 的強(弱)度值: [-0.6540457576770174, -0.44919510789257716, 0.4510310748149666, 0.5083996944782051, 0.5481638272347492, 0.48182915162232764, 0.438679954560916, 0.0896474561929245, 0.05064336630259022, 0.28427510812751017, 0.534818635338587, 0.33076879867050446, 0.5820236955136912, 0.6486044011470377, 0.43562538678146234, 0.33038816122302384, 0.08868984250851532, -0.08936748323870063, -0.2671964225043414, 0.07935427751242484, 0.34253085061406646, -0.02947064046825129, 0.11518412710836297, -0.060914158621582296, 0.5228250277495355, 0.3858176806739346, 0.26186260393425803, 0.44334945881476856, 0.4294817907186277, 0.6386445068563965, 0.39398816944346365, 0.5567565226648957, -0.42588945023043456, -0.11341267808539206, 0.5094281820792802, 0.6387262990939764, 0.5942874129393058, 0.40269242724946797, 0.8327447608662222, 0.7379108870214749, 0.5482441759172

In [10]:
for i, group_index in enumerate(range(len(twii_close) - window_size + 1)):
    twii_start_price = twii_close.iloc[group_index]
    twii_end_price = twii_close.iloc[group_index + window_size - 1]
    twii_20days_return = ((twii_end_price - twii_start_price) / twii_start_price) * 100

    t2300_start_price = t2300_close.iloc[group_index]
    t2300_end_price = t2300_close.iloc[group_index + window_size - 1]
    t2300_20days_return = ((t2300_end_price - t2300_start_price) / t2300_start_price) * 100

    strength_value = (t2300_20days_return - twii_20days_return) / twii_20days_return

    print(f"第 {i+1} 組第 19 筆台股加權指數: {twii_close.index[group_index].date()}, 收盤價: {twii_close.iloc[group_index]}, 20 天的漲跌幅%: {twii_20days_return:.2f}, 台積電20天的漲跌幅%: {t2300_20days_return:.2f}, 台積電(2330) 的強(弱)度值: {strength_value:.2f}")


第 1 組第 19 筆台股加權指數: 2019-01-02, 收盤價: 9554.1396484375, 20 天的漲跌幅%: 3.95, 台積電20天的漲跌幅%: 1.37, 台積電(2330) 的強(弱)度值: -0.65
第 2 組第 19 筆台股加權指數: 2019-01-03, 收盤價: 9492.419921875, 20 天的漲跌幅%: 4.63, 台積電20天的漲跌幅%: 2.55, 台積電(2330) 的強(弱)度值: -0.45
第 3 組第 19 筆台股加權指數: 2019-01-04, 收盤價: 9382.509765625, 20 天的漲跌幅%: 6.63, 台積電20天的漲跌幅%: 9.62, 台積電(2330) 的強(弱)度值: 0.45
第 4 組第 19 筆台股加權指數: 2019-01-07, 收盤價: 9590.2998046875, 20 天的漲跌幅%: 5.29, 台積電20天的漲跌幅%: 7.98, 台積電(2330) 的強(弱)度值: 0.51
第 5 組第 19 筆台股加權指數: 2019-01-08, 收盤價: 9563.599609375, 20 天的漲跌幅%: 5.51, 台積電20天的漲跌幅%: 8.53, 台積電(2330) 的強(弱)度值: 0.55
第 6 組第 19 筆台股加權指數: 2019-01-09, 收盤價: 9738.3095703125, 20 天的漲跌幅%: 3.60, 台積電20天的漲跌幅%: 5.34, 台積電(2330) 的強(弱)度值: 0.48
第 7 組第 19 筆台股加權指數: 2019-01-10, 收盤價: 9720.6904296875, 20 天的漲跌幅%: 3.54, 台積電20天的漲跌幅%: 5.09, 台積電(2330) 的強(弱)度值: 0.44
第 8 組第 19 筆台股加權指數: 2019-01-11, 收盤價: 9759.400390625, 20 天的漲跌幅%: 3.95, 台積電20天的漲跌幅%: 4.31, 台積電(2330) 的強(弱)度值: 0.09
第 9 組第 19 筆台股加權指數: 2019-01-14, 收盤價: 9708.2197265625, 20 天的漲跌幅%: 4.57, 台積電20天的漲跌幅%: 4.81, 台積電(2330)

## 弱勢股舉例


晶電(2448)20天的強(弱)度值
步驟一、計算出台股加權指數20天的漲跌幅%是+ 1.85%
前面已經計算過故不再贅述。
步驟二、再算出晶電(2448)20天的漲跌幅%
2019/3/4 晶電收盤價是27.05元，2019/3/29晶電收盤價是24.8元
晶電20天的漲跌幅%=((24.8-27.05)/27.05)x100= - 8.32%
步驟三、計算晶電強(弱)度值
( - 8.32 –1.85)/1.85= - 5.5
由以上實例，我們可以算出晶電(2448)的強(弱)度值是 – 5.497

所以很明顯的比較出台積電(2330)是強勢股而晶電(2448)是弱勢股
這裡面的20天是中期強度，若要計算短期強度則可以把20天改為5天，
若要計算長期強度則可以把20天改為120天計算。
